### ISSUE TITLE
Handle package directories atomically

### ISSUE THREAD

**Issue by didenko** [sahib/rmlint#394]
OSX-style documents are packages - in that they are specially-named directories which contain a structure of files. Changing only some files inside such a package has a high chance to break the document integrity so that the software which made the document will consider it corrupt.

Not only Apple's `*.pages` and similar documents are affected. There are third-party packages affected - from likes of `*.li3d` and `*.screenflow`, to Adobe Lightroom's and other photo management catalogs - the latter has a cross-platform impact as it is often shared between systems by moving DAS. While it has extensions for some directories (`*.lrcat`, `*.lrdata`), some other directories are to be distinguished by "magic" content:

```bash
[vlad@box Lightroom]$ ls
 Backups
'Download Backups'
'Lightroom Catalog.lrcat'
'Lightroom Catalog Previews.lrdata'
'Lightroom Catalog Smart Previews.lrdata'
'Lightroom Settings'
[vlad@box Lightroom]$ ls 'Lightroom Settings'
'Auto Layout Presets'      'Local Adjustment Presets'
'Develop Presets'           Locations
'Export Presets'           'Metadata Presets'
'External Editor Presets'  'Print Templates'
'Filename Templates'       'Slideshow Templates'
'Filter Presets'           'Text Style Presets'
'FTP Presets'              'Text Templates'
'Import Presets'           'Video Presets'
'Keyword Sets'              Watermarks
'Label Sets'               'Web Templates'
[vlad@box Lightroom]$ 
```
The suggestion is to introduce a package-handling behavior and a related configuration parameter, e.g.:

> --as-package: treat the matching directories as atomic packages, meaning that if and only is all contained objects are identical to a tagged destination's, then a matching directory is considered a duplicate.
> parameter's value examples:
>    --as-package path:'/mnt/media/Photo/Lightroom Settings'
>    --as-package ext:lrcat,lrdata,li3d
>    --as-package child:'Local Adjustment Presets'

**Comment by sahib** [comment]
Hello @didenko,

I see the issue, but I do not agree with the proposed solution. Your problem seems to be that you want to keep the integrity of those directory - i.e. other programs should not notice any difference before and after deduplication. This sounds a lot like a case for using hardlinks. So instead of removing those files, you replace them with a hardlink to the original:

```bash
$ rmlint -c sh:link
```

> The suggestion is to introduce a package-handling behavior and a related configuration parameter, e.g.:

This is a very complex proposal (i.e. lots of code I need to write & maintain) for a very specific problem (i.e. very few users). In other words: I do not think the added complexity is outweighed by the usability this adds.   

> Not only Apple's *.pages and similar documents are affected. There are third-party packages affected - from likes of *.li3d and *.screenflow, to Adobe Lightroom's and other photo management catalogs - the latter has a cross-platform impact as it is often shared between systems by moving DAS. While it has extensions for some directories (*.lrcat, *.lrdata), some other directories are to be distinguished by "magic" content:

For those kind of specific use-cases we built the ``json`` output. This makes scripting own things relatively easy if you need to:

```bash
# Example: List all original files, one per line:
$ rmlint -o json  | jq -r '.[] | select(.is_original) | .path'
```

**Comment by misieck** [comment]
I might add that if all you want is to remove duplicate "package" directories, you could use
`rmlint -T dd`
This will only search for whole duplicate directories and leave loose files alone.

**Comment by didenko** [comment]
@sahib Thank you for looking into this.

The hard links suggestion will be troublesome to rid of duplicates across filesystems for my use case. `rmlint` turned out to be great to converge back trees which users diverged some time ago - I gradually bring them back to one shared filesystem.

`rmlint` has a really good workflow to deal with this - separating the discovery, validation, and actual cleanup. Thank you for the tool!

I am no the one to propose a solution - my suggestion was more to illustrate where (in spirit) I would like to get, not necessarily the method. The issue affects a set of OSX users using the iWork (or whatever the current suite's name is) and almost any other OSX application that saves documents as directories. I am not seeing how this is very few users (as much as I would love it to be a small set :) )

I will play to see if I can script around the `jq` output - that may indeed be the way forward - appreciate the query line!

@misieck Thank you, will play with `-T dd` again. I did before and did not get the effect I expected - seemed not to detect directory-documents as duplicates deep in a tree, which had subdirs themselves. I will check it again.

**Comment by sahib** [comment]
Hello @didenko,

> The hard links suggestion will be troublesome to rid of duplicates across filesystems for my use case. rmlint turned out to be great to converge back trees which users diverged some time ago - I gradually bring them back to one shared filesystem.

The option ``-c sh:link`` is actually clever enough to create symbolic links if the files are on different disks. Maybe that's helpful to you.

> I am no the one to propose a solution - my suggestion was more to illustrate where (in spirit) I would like to get, not necessarily the method. The issue affects a set of OSX users using the iWork (or whatever the current suite's name is) and almost any other OSX application that saves documents as directories. I am not seeing how this is very few users (as much as I would love it to be a small set :) )

My problem is not that the number of users, but the balance between how specific the feature (quite specific) and the amount of complexity it would add to the code base (high). If I would allow this kind of features to be added, there would be a lot of other use cases that would be equally justified to be implemented. ``rmlint`` already has too many options, so I'm quite conservative to this and rather concentrate on improving the current features.

> I will play to see if I can script around the jq output - that may indeed be the way forward - appreciate the query line!

I should note that this line is only an arbitrary example, but together with the ``jq`` docs it should get you going. 

> @misieck Thank you, will play with -T dd again. I did before and did not get the effect I expected - seemed not to detect directory-documents as duplicates deep in a tree, which had subdirs themselves. I will check it again.

Hmm, there were some issues which are partly fixed (listed in #385). Those will be hopefully better in ``2.10.0`` (i.e. the next version).

----

Good luck. :smirk:
